apps/dapp-console-api/src/middleware/faucetClaimRateLimiter.ts

```diff
@@ -0,0 +1,45 @@
+import { TRPCError } from '@trpc/server'
+import type { Logger } from 'pino'
+
+import { envVars } from '@/constants'
+import type { GrowthbookStore } from '@/growthbook'
+import { metrics } from '@/monitoring/metrics'
+import type { Trpc } from '@/Trpc'
+import { getFaucetClaimRate } from '@/utils'
+import type { RedisCache } from '@/utils/redis'
+
+/** Middleware used for checking if a user is under the faucet claim rate limit */
+export const faucetClaimRateLimiter = (
+  trpc: Trpc,
+  growthBookStore: GrowthbookStore,
+  redisCache: RedisCache,
+  logger?: Logger,
+) => {
+  return trpc.middleware(async ({ ctx, next }) => {
+    const { ip } = ctx.req
+    console.log(
+      'enable_faucet_rate_limit:',
+      growthBookStore.get('enable_faucet_rate_limit'),
+    )
+    if (!growthBookStore.get('enable_faucet_rate_limit')) {
+      return next()
+    }
+
+    if (!ip) {
+      logger?.warn('faucet encountered a request with an undefined ip address')
+      return next()
+    }
+
+    const requestCount = await getFaucetClaimRate(redisCache, ip)
+    if (requestCount && requestCount >= BigInt(envVars.FAUCET_RATE_LIMIT)) {
+      logger?.warn('faucet request rate limited')
```

Comment:
Lets also log the ip address that is rate limited so that we have an idea of what's being limited